File: go/wasip2/conversion.go
package wasi_otel

import (
	"time"

	wasiTrace "github.com/calebschoepp/opentelemetry-wasi/internal/wasi/otel/tracing"
	"go.bytecodealliance.org/cm"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/sdk/instrumentation"
	sdkTrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"
)

func toWasiTraceState(s trace.TraceState) wasiTrace.TraceState {
	// TODO: I don't trust that this is the proper way to create/add to a cm.List;
	// it seems like it would just overwrite the existing list with a single, latest entry
	result := cm.NewList[[2]string](nil, 0)
	s.Walk(func(k, v string) bool {
		result = cm.NewList(&[2]string{k, v}, 2)
		return true
	})

	return wasiTrace.TraceState(result)
}

func toWasiDateTime(t time.Time) wasiTrace.DateTime {
	// TODO: Make sure this is retrieving the correct time
	return wasiTrace.DateTime{
		Seconds:     uint64(t.Second()),
		Nanoseconds: uint32(t.Nanosecond()),
	}
}

func toWasiAttributes(attributes []attribute.KeyValue) cm.List[wasiTrace.KeyValue] {
	result := cm.NewList[wasiTrace.KeyValue](nil, 0)
	for _, entry := range attributes {
		result = cm.NewList(&wasiTrace.KeyValue{
			Key:   wasiTrace.Key(entry.Key),
			Value: toWasiValue(entry.Value),
		}, len(attributes))
	}

	return result
}

func toWasiEvents(events []sdkTrace.Event) cm.List[wasiTrace.Event] {
	result := cm.NewList[wasiTrace.Event](nil, 0)
	for _, event := range events {
		result = cm.NewList(&wasiTrace.Event{
			Name:       event.Name,
			Time:       toWasiDateTime(event.Time),
			Attributes: toWasiAttributes(event.Attributes),
		}, len(events))
	}

	return result
}

func toWasiSpanContext(s trace.SpanContext) wasiTrace.SpanContext {
	return wasiTrace.SpanContext{
		TraceID:    wasiTrace.TraceID(s.TraceID().String()),
		SpanID:     wasiTrace.SpanID(s.SpanID().String()),
		TraceFlags: wasiTrace.TraceFlags(s.TraceFlags()),
		IsRemote:   s.IsRemote(),
		TraceState: toWasiTraceState(s.TraceState()),
	}
}

func toWasiLinks(links []sdkTrace.Link) cm.List[wasiTrace.Link] {
	result := cm.NewList[wasiTrace.Link](nil, 0)
	for _, link := range links {
		result = cm.NewList(&wasiTrace.Link{
			SpanContext: toWasiSpanContext(link.SpanContext),
			Attributes:  toWasiAttributes(link.Attributes),
		}, len(links))
	}

	return result
}

func toWasiInstrumentationScope(s instrumentation.Scope) wasiTrace.InstrumentationScope {
	return wasiTrace.InstrumentationScope{
		Name:       s.Name,
		Version:    getOption(s.Version),
		SchemaURL:  getOption(s.SchemaURL),
		Attributes: toWasiAttributes(s.Attributes.ToSlice()),
	}
}

func toWasiStatus(s sdkTrace.Status) wasiTrace.Status {
	switch s.Code {
	case codes.Error:
		return wasiTrace.StatusError(s.Description)
	case codes.Ok:
		return wasiTrace.StatusOK()
	default:
		return wasiTrace.StatusUnset()
	}
}

func toWasiValue(v attribute.Value) wasiTrace.Value {
	switch v.Type() {
	case attribute.STRING:
		return wasiTrace.ValueString_(v.AsString())
	case attribute.BOOL:
		return wasiTrace.ValueBool(v.AsBool())
	case attribute.FLOAT64:
		return wasiTrace.ValueF64(v.AsFloat64())
	case attribute.INT64:
		return wasiTrace.ValueS64(v.AsInt64())
	case attribute.STRINGSLICE:
		stringSlice := v.AsStringSlice()
		return wasiTrace.ValueStringArray(cm.ToList(stringSlice))
	case attribute.BOOLSLICE:
		boolSlice := v.AsBoolSlice()
		return wasiTrace.ValueBoolArray(cm.ToList(boolSlice))
	case attribute.FLOAT64SLICE:
		floatSlice := v.AsFloat64Slice()
		return wasiTrace.ValueF64Array(cm.ToList(floatSlice))
	case attribute.INT64SLICE:
		intSlice := v.AsInt64Slice()
		return wasiTrace.ValueS64Array(cm.ToList(intSlice))
	default:
		return wasiTrace.ValueString_(v.AsString())
	}
}

func getOption[T comparable](v T) cm.Option[T] {
	// TODO: Claude claims that this is the best way to do this function. VALIDATE
	var empty T
	if v == empty {
		return cm.None[T]()
	}
	return cm.Some(v)
}

File: rust/src/tracing/conversion.rs
use crate::wit::wasi::otel::tracing::*;

impl From<opentelemetry_sdk::trace::SpanData> for SpanData {
    fn from(value: opentelemetry_sdk::trace::SpanData) -> Self {
        Self {
            span_context: value.span_context.into(),
            parent_span_id: value.parent_span_id.to_string(),
            span_kind: value.span_kind.into(),
            name: value.name.to_string(),
            start_time: value.start_time.into(),
            end_time: value.end_time.into(),
            attributes: value.attributes.into_iter().map(|e| (&e).into()).collect(),
            events: value.events.events.into_iter().map(Into::into).collect(),
            links: value.links.links.into_iter().map(Into::into).collect(),
            status: value.status.into(),
            instrumentation_scope: (&value.instrumentation_scope).into(),
            dropped_attributes: value.dropped_attributes_count,
            dropped_events: value.events.dropped_count,
            dropped_links: value.links.dropped_count,
        }
    }
}

impl From<opentelemetry::trace::SpanContext> for SpanContext {
    fn from(value: opentelemetry::trace::SpanContext) -> Self {
        Self {
            trace_id: format!("{:x}", value.trace_id()),
            span_id: format!("{:x}", value.span_id()),
            trace_flags: value.trace_flags().into(),
            is_remote: value.is_remote(),
            trace_state: value
                .trace_state()
                .header()
                .split(',')
                .filter_map(|s| {
                    if let Some((key, value)) = s.split_once('=') {
                        Some((key.to_string(), value.to_string()))
                    } else {
                        None
                    }
                })
                .collect(),
        }
    }
}

impl From<SpanContext> for opentelemetry::trace::SpanContext {
    fn from(value: SpanContext) -> Self {
        let trace_id = opentelemetry::trace::TraceId::from_hex(&value.trace_id)
            .unwrap_or(opentelemetry::trace::TraceId::INVALID);
        let span_id = opentelemetry::trace::SpanId::from_hex(&value.span_id)
            .unwrap_or(opentelemetry::trace::SpanId::INVALID);
        let trace_state = opentelemetry::trace::TraceState::from_key_value(value.trace_state)
            .unwrap_or_else(|_| opentelemetry::trace::TraceState::default());
        Self::new(
            trace_id,
            span_id,
            value.trace_flags.into(),
            value.is_remote,
            trace_state,
        )
    }
}

impl From<opentelemetry::trace::TraceFlags> for TraceFlags {
    fn from(value: opentelemetry::trace::TraceFlags) -> Self {
        if value.is_sampled() {
            TraceFlags::SAMPLED
        } else {
            TraceFlags::empty()
        }
    }
}

impl From<TraceFlags> for opentelemetry::trace::TraceFlags {
    fn from(value: TraceFlags) -> Self {
        Self::new(value.bits())
    }
}

impl From<opentelemetry::trace::SpanKind> for SpanKind {
    fn from(value: opentelemetry::trace::SpanKind) -> Self {
        match value {
            opentelemetry::trace::SpanKind::Client => Self::Client,
            opentelemetry::trace::SpanKind::Server => Self::Server,
            opentelemetry::trace::SpanKind::Producer => Self::Producer,
            opentelemetry::trace::SpanKind::Consumer => Self::Consumer,
            opentelemetry::trace::SpanKind::Internal => Self::Internal,
        }
    }
}

impl From<opentelemetry::trace::Event> for Event {
    fn from(value: opentelemetry::trace::Event) -> Self {
        Self {
            name: value.name.to_string(),
            time: value.timestamp.into(),
            attributes: value.attributes.into_iter().map(|e| (&e).into()).collect(),
        }
    }
}

impl From<opentelemetry::trace::Link> for Link {
    fn from(value: opentelemetry::trace::Link) -> Self {
        Self {
            span_context: value.span_context.into(),
            attributes: value.attributes.into_iter().map(|e| (&e).into()).collect(),
        }
    }
}

impl From<opentelemetry::trace::Status> for Status {
    fn from(value: opentelemetry::trace::Status) -> Self {
        match value {
            opentelemetry::trace::Status::Unset => Self::Unset,
            opentelemetry::trace::Status::Error { description } => {
                Self::Error(description.to_string())
            }
            opentelemetry::trace::Status::Ok => Self::Ok,
        }
    }
}
